lib/prompb/prompb.go

```diff
@@ -102,57 +227,93 @@ func (wr *WriteRequest) UnmarshalProtobuf(src []byte) (err error) {
 	wr.Timeseries = tss
 	wr.labelsPool = labelsPool
 	wr.samplesPool = samplesPool
+	wr.exemplarsPool = exemplarsPool
+	wr.histogramsPool = histogramsPool
 	return nil
 }
 
-func (ts *TimeSeries) unmarshalProtobuf(src []byte, labelsPool []Label, samplesPool []Sample) ([]Label, []Sample, error) {
+func (ts *TimeSeries) unmarshalProtobuf(src []byte, labelsPool *[]Label, samplesPool *[]Sample, exemplarsPool *[]Exemplar, histogramsPool *[]Histogram) error {
 	// message TimeSeries {
-	//   repeated Label labels   = 1;
-	//   repeated Sample samples = 2;
+	//   repeated Label labels         = 1;
+	//   repeated Sample samples       = 2;
+	//   repeated Exemplar exemplars   = 3;
+	//   repeated Histogram histograms = 4;
 	// }
-	labelsPoolLen := len(labelsPool)
-	samplesPoolLen := len(samplesPool)
+	labelsPoolLen := len(*labelsPool)
+	samplesPoolLen := len(*samplesPool)
+	exemplarsPoolLen := len(*exemplarsPool)
+	histogramsPoolLen := len(*histogramsPool)
 	var fc easyproto.FieldContext
 	for len(src) > 0 {
 		var err error
 		src, err = fc.NextField(src)
 		if err != nil {
-			return labelsPool, samplesPool, fmt.Errorf("cannot read the next field: %w", err)
+			return fmt.Errorf("cannot read the next field: %w", err)
 		}
 		switch fc.FieldNum {
 		case 1:
 			data, ok := fc.MessageData()
 			if !ok {
-				return labelsPool, samplesPool, fmt.Errorf("cannot read label data")
+				return fmt.Errorf("cannot read label data")
 			}
-			if len(labelsPool) < cap(labelsPool) {
-				labelsPool = labelsPool[:len(labelsPool)+1]
+			if len(*labelsPool) < cap(*labelsPool) {
```

Comment:
It's possible to replace copy paste with generic function:

```go
func advanceLenMayOverAllocate[T any](dst []T) []T {
	if len(dst) < cap(dst) {
		dst = dst[:len(dst)+1]
	} else {
		var d T
		dst = append(dst, d)
	}
	return dst
}
```

usage: 

```go
sp.labelsPool = advanceLenMayOverAllocate(sp.labelsPool)
label := labelsPool[len(lsp.abelsPool)-1]
```

Reply:
replaced it with slices.Grow + append

Reply:
updated